File: Cpp/0041_first_missing_positive.cpp
// https://leetcode.cn/problems/first-missing-positive/

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            int num = nums[i];
            while ((num >= 1) && (num <= n)) {
                int tmp = nums[num-1];
                if (tmp == num) {
                    break;
                }

                nums[num-1] = num;
                nums[i] = tmp;
                num = tmp;
            }
        }

        for (int i = 0; i < n; ++i) {
            if (nums[i] != i+1) {
                return i+1;
            }
        }
        return n+1;
    }
};

File: Python3/0041_first_missing_positive.py
# https://leetcode.cn/problems/first-missing-positive/

class Solution:
    def firstMissingPositive(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: int
        """

        n = len(nums)
        for i in range(n):
            num = nums[i]
            while (num >= 1) and (num <= n):
                tmp = nums[num-1]
                # If the (num-1)th number is already num, break.
                if tmp == num:
                    break
                    
                # Swap nums[num-1] and nums[i].
                nums[num-1] = num
                nums[i] = tmp
                num = tmp

        for i, num in enumerate(nums):
            if i != num - 1:
                return i+1
        return n+1